src/menus/PluginMenus.cpp

```diff
@@ -32,6 +32,107 @@
 // private helper classes and functions
 namespace {
 
+using MenuList = std::vector<std::pair<TranslatableString, std::vector<TranslatableString>>>;
+
+const MenuList& PreferredEffectsMenu()
+{
+   static MenuList items =
+   {
+      {
+         XO("Volume and compression"),
+         {
+            XO("Amplify"),
```

Comment:
I don't like the closed-endedness -- the violation of the "O", open-closed principle, of SOLID:  "A design should be open to extension but closed to modification"

In other words, if we ever add a new effect that falls into one of these categories, then it won't be self-contained in one new implementation file, but instead this file will need modification to make the menu sorting continue to work as desired.

And we are also depending on long-distance coincidences of string literal values, between this file and each of the many effect implementation files, which is хрупкость.

So I suggest:

- A new enumeration EffectCategory, defined in Effect.h
- A table mapping EffectCategory to EnumValueSymbol (which pairs an identifier and TranslatableString), defined in Effect.cpp and accessible by a static member function
- MenuList is instead `std::vector<std::pair<EffectCategory, std::vector<TranslatableString>>>` and gets TranslatableString from that table
- A virtual function of Effect called GetEffectCategory, appropriately overridden in all the effects that matter
- Some of these effects are not built-ins, but Nyquist plug-ins. So NyquistEffect needs an override of GetEffectCategory that doesn't just return a constant, but computes an answer too.
  - Add a case in NyquistEffect::Parse to recognize a new special Lisp comment `$category`, which does reverse-lookup of the table of categories, by the Identifier (not by the TranslatableString), to find the enum value
  - Modify the plug-ins/*.ny files as needed to identify their categories
  - Ping @SteveDaulton when you do this, to review the changes at least in Nyquist.cpp, and recommend how to update documentation of Nyquist
- PreferredEffectsMenu does not return a static value, but computes its answer instead, by iterating the plug-ins and querying GetEffectCategory
- Maybe it can cache that computed answer, but invalidation of the cache would need to be done by the plugin manager dialog, or if the user changes the Language preference; see class PreferenceListener

Do all of this, and I will be less dissatisfied... but then the set of categories and the sorting of the categories themselves will still be closed-ended, and in a further future I might want to seek a fix for that too involving a registration of categories.

Reply:
1. EffectCategory is very likely to be extended/changed.
2. Not each effect has should have effect category assigned, effects can be added or removed from that list.
3. MenuList can be changed at any time, without modification of the effects source code, it's a resource object.
Considering that, I don't agree that my implementation violates "O"

Reply:
Your implementation violates open ended ness for reasons I described.  You must change PluginMenus.cpp if in future a new categorized effect is added, rather than having all necessary new code self contained in the effect’s implementation.  I do not accept this.

Reply:
> 1. EffectCategory is very likely to be extended/changed.

How does that make me wrong?

It is something I admitted: that even if you do my steps, it is not a complete fix for the problem of closed ended ness.  Still the steps I described are necessary if not sufficient.

> 2. Not each effect should have effect category assigned, effects can be added or removed from that list.

Do you mean there should be another enumeration value for effects of undefined category?  That is easy enough.

“Effects can be added or removed from the list.”  Do you mean as future development, or do you mean that the composition of the lists can be changed at runtime?

> 3. MenuList can be changed at any time, without modification of the effects source code, it's a resource object.
> 

Do you mean that as a future development, we change our decision about the categorization of an individual effect?  But it is much more proper, if some source code must change, that title is the source code of that effect only rather than PluginMenus.cpp.

> Considering that, I don't agree that my implementation violates "O". Ideally, there could be an external file describing the stucture of that menu, which can be changed without any code modification.

Reply:
"MenuList can be changed at any time, without modification of the effects source code, it's a resource object.
Does “resource object” mean that the categorization of individual effects should not be hard oder in them, but somehow specified in an external file, like the configuration, and the user would have some way to change its contents, and the program would do effect sorting according to an interpretation of that file? But this implementation does none of that."

So I do t understand what you mean by “resource object.”

Reply:
Sorry I edited your original comment by mistake.  I didn't mean to delete it.  But I think I quoted all of it in my replies.

Reply:
"Maybe it can cache that computed answer, but invalidation of the cache would need to be done by the plugin manager dialog, or if the user changes the Language preference; see class PreferenceListener"

To clarify I mention this really to suggest you don't bother with it:  Do recalculate the list each time you need to repopulate the effects menus.

Reply:
> > How does that make me wrong?
> 
> "A design should be open to extension but closed to modification"

That doesn't answer my question.

> 
> > it is not a complete fix for the problem of closed ended ness
> 
> Please, let do not pollute interfaces with more constants that has some meaning for about 30 effects, it has nothing to do with OC.

What "interface" is polluted?  Effect itself?  Well it's going through lots of necessary changes this release.  But it is quite simple to make effects un-categorized with a default implementation of the virtual function, and so most effects will need no modification for this.  Only the few that are categorized.

> 
> > Do you mean there should be another enumeration value for effects of undefined category?
> 
> Why should prebundled effect, which isn't present in that menu, should be categorized as "Undefined"? Or why should any other external effect have category which will never be used?

This really isn't different in principle than IsHiddenFromMenus.

> 
> > Do you mean as future development, or do you mean that the composition of the lists can be changed at runtime?
> 
> During development ofcourse, and this is in fact points to the possibility to violate OC in future.

We are failing to understand "OC" to mean the same thing.

If there is a new effect, its categorization should be determined by the code of that effect alone.

The file that makes effect menus should be generalized, and should not make special mentions of any individual effects.  The interface offered by abstract Effect is extended now to be sufficient for the needs of the plugin menus.

Do you see there is special weird stuff in PluginMenus about noise reduction and the Nyquist prompt?  True.  But that is bad legacy, not good precedent.

> 
> > Do you mean that as a future development, we change our decision about the categorization of an individual effect? But it is much more proper, if some source code must change, that is the source code of that effect only rather than PluginMenus.cpp.
> 
> Yes, we can add or remove effects to/from that list, even now there are prebundled effects that are not present in that list, the set of groups can change as well. Question: there are groups like "Distortion and Modulation", "EQ and Filters", should there be 4 emum values or only 2? What if effect provides both Distortion and EQ? It's not a problem from the implementation point of view, and it's not a bad idea in general, but there is no use for it right now.

As I said, I am admitting to incompleteness of this solution for open endedness, but I do not accept your argument that therefore, let this PR stand and make no attempt to fix it.

Effects should have some virtual function reporting their category as some kind of value, and PluginMenus.cpp should use that value and not have a hard-coded list of effect names.

Start by making that value an enumeration.  Though this enumeration of categories is closed ended, the set of effects in any one category is not closed ended.

Let me see the redraft.  Then we can discuss the further step of changing the value type from enumeration to some other dynamically determined set of cateogires.

> 
> > Does “resource object” mean that the categorization of individual effects should not be hard oder in them, but somehow specified in an external file, like the configuration, and the user would have some way to change its contents, and the program would do effect sorting according to an interpretation of that file? But this implementation does none of that.
> 
> There is a set of effects that should be put in a certaing menu groups, there is no need to "categorize" all effects. MenuLists can be put into separate cpp file (like AllThemeResources), or even external xml (json, txt,...) file. Not sure that we need to provide a way for user to change that list, but surely we can do that. In fact we can extend format with rules and filters and provide a way to make this menus very customizable, and your idea with effects category will make it more flexible for the user.

Reply:
>That doesn't answer my question.

What question then? Using enum for entity like effect category is very likely to violate OC (until we agree that it will never change). So I don't see why your impementation violating OC is preferrable over my implementation which "violates OC" (which is not, because you don't have to change effect to change menu ordering, and you don't have to change anything in the menu to add a new effect, though you may want to change that too, but the problem here is that you cannot answer to the question "how should I change that menu?")

>If there is a new effect, its categorization should be determined by the code of that effect alone.

Effect placement in the menu should be determined by the menu. It can use properties of the effect to do that and using effect ID for that purpose isn't worse than using any other property of the effect. 

>The file that makes effect menus should be generalized, and should not make special mentions of any individual effects.

Effect ID and effect category are both properties of the same entity, there is no difference between them in sence that they both could to be used as a filter parameters.

This task does not specify that we should assign category to each effect, and build menu so that all effects grouped by these categories. This task requires to put **certain** effects into certain submenus, that's why there are references to the individual effects in the menu description, you cannot solve such task using generalization.

Reply:
> > That doesn't answer my question.
> 
> What question then? Using enum for entity like effect category is very likely to violate OC (until we agree that it will never change). So I don't see why your impementation violating OC is preferrable over my implementation which "violates OC" (which is not, because you don't have to change effect to change menu ordering, and you don't have to change anything in the menu to add a new effect, though you may want to change that too, but the problem here is that you cannot answer to the question "how should I change that menu?")

One more time:  I admit the incompleteness of my suggestions, but still say it is a better organization than this first attempt.  If you are actually arguing to go beyond my suggestions and figure out open-endedness of categories, not just the members of categories -- great, do so!

If not, still I want some of the extra effort done now.

But the rest of what you say is unpersuasive to me.  Saying you only satisfy the issue in a looks-like-it-works way in the black box view is not acceptable.

The decision to categorize an individual effect one way or another way should be encoded in the implementation of that effect.  That is the open-ended solution in regard to the enlargement of the set of effects.

Ideally effects should be plug-ins, not baked into the executable.  In fact this is realized in the case of Nyquist effects, though we don't do this yet for effects in C++.

At least for Nyquist effects, a Nyquist programmer should have the ability to put a new effect into the sub-menu of their choice, just by writing appropriate code in the .ny file, without needing special cooperation from a re-compiled Audacity executable.

Reply:
>but still say it is a better organization

I say that they are orthogonal to each other.

>looks-like-it-works

What exactly does not work here?

>At least for Nyquist effects, a Nyquist programmer should have the ability to put a new effect into the sub-menu of their choice, just by writing appropriate code in the .ny file, without needing special cooperation from a re-compiled Audacity executable.

It's opposite to what is specified in the task:
>[any external plugins, whether it be nyquist, VST or LV2, arranged as thus:]
Author/Publisher
Effect, sorted alphabetically by name

Reply:
> > but still say it is a better organization
> 
> I say that they are orthogonal to each other.

I don't understand what is orthogonal to what.

> 
> > looks-like-it-works
> 
> What exactly does not work here?

Nothing.  BUT what I say from point of view of architecture is, it's not enough that it merely looks like it works.  It shoudl be properly designed for extension.

> 
> > At least for Nyquist effects, a Nyquist programmer should have the ability to put a new effect into the sub-menu of their choice, just by writing appropriate code in the .ny file, without needing special cooperation from a re-compiled Audacity executable.
> 
> It's opposite to what is specified in the task:
> 
> > [any external plugins, whether it be nyquist, VST or LV2, arranged as thus:]
> > Author/Publisher
> > Effect, sorted alphabetically by name

Then @LWinterberg should respecify the issue.

We can specify the interface of Nyquist programming as we can't for the other protocols.  We can and should make the effort to give Nyquist programmers the ability to categorize effects as they choose.

Reply:
I think @SteveDaulton would agree with the last point.